fix: update tab with new attributes

smbea opened the pull request:
Closes #2581

Explanation: the spread operator only keeps the enumerable properties of the object itself. Therefore, functions like `getTitle()` were getting lost.

MaxTru reviewed (commented):
Great!

Maybe use our utility function for `assign`?
https://github.com/camunda/camunda-modeler/pull/2649/commits/e92bf615c55d7abd86874b846bade995a26919bd#diff-b650ad882ccc34a87aa822cb69b9bffc2e3c1513db7dbdff0fc8aaec11fd99a2L16

Otherwise we have two different `assign` usage approaches even within the same source code file.

barmac commented:
Can you please provide a test case so that we catch the bug if it reappears?

smbea commented:
I added new changes because, as it turns out, that using `assign` still causes the loss of the getters, when trying to clone a tab object. I added a `cloneWithDescriptors` function which clones an object, keeping the getters/setters of the properties. Once this is done, then we can assign the new attributes:

```javascript 
let updatedTab = cloneWithDescriptors(tab);
assign(updatedTab, newAttrs);
```

I also added a test case to ensure this stays functional.

This was the simplest solution I found for this issue. Let me know what you think.

MaxTru commented:
LGTM. I will not approve because I want to give @barmac opporunity to review as well

MaxTru commented on client/src/app/__tests__/AppSpec.js:
```diff
@@ -2177,7 +2193,6 @@ describe('<App>', function() {
       } = app.state;
 
       // then
-      expect(updatedTab).to.not.eql(tab);
```
Why is this needed? Does this even make sense?

smbea replied:
I missed the line that made this make sense. Here it is:
 
 ```diff

-  expect(updatedTab).to.not.eql(tab);
+  expect(updatedTab).to.include(newAttrs);

```
This was causing the test to fail. Even with the new function I added, the old tab object is updated with the new properties. I'm not sure how we can prevent this from happening easily.

I didn't consider assertion relevant, because the "navigation history" test passes, and the new tab is updated correctly (as proved by the new line added).

But let me know if you have different insight on this.

barmac commented on client/src/app/__tests__/AppSpec.js:
```diff
@@ -2139,6 +2139,22 @@ describe('<App>', function() {
+      // when
+      const updatedTab = await app.triggerAction('lint-tab', { tab });
+
+      const nameGetter = Object.getOwnPropertyDescriptor(updatedTab, 'name').get;
+      const titleGetter = Object.getOwnPropertyDescriptor(updatedTab, 'name').get;
```
This is super low-level. I'd rather expect just a check if the tab name and title are correct (value test).

barmac commented:
I was just wondering... How about if we use [`TabsProvider#createForFile`](https://github.com/camunda/camunda-modeler/blob/develop/client/src/app/TabsProvider.js#L461) to create a new tab from the same file and then just make sure we properly clone the ID and update the properties? This way we could perhaps avoid the problem of cloning the descriptors.

smbea commented:
> I was just wondering... How about if we use TabsProvider#createForFile to create a new tab from the same file and then just make sure we properly clone the ID and update the properties? This way we could perhaps avoid the problem of cloning the descriptors.

The reason for that is because that would generate a different ID, and would not work with this:

```javascript
// replace activeTab
let updatedActiveTab = activeTab;
if (activeTab.id === updatedTab.id) {
    updatedActiveTab = updatedTab;
}
```

barmac commented:
True, but we can just keep the old id, right?
```javascript
const newTab = tabsProvider.createTabForFile(oldTab.file);
newTab.id = oldTab.id;
```

smbea commented:
That works and ridiculously simpler 😂

barmac commented:
https://user-images.githubusercontent.com/28307541/147125666-af38e5d2-d5cf-4118-98b9-f1fec6316c52.mov

Hmm the new solution introduces a new bug which is uncaught on the CI :/

barmac commented:
(My comment was about the state before your most recent push).

smbea commented:
> (My comment was about the state before your most recent push).

I only simplified the test case in the most recent push

smbea commented:
I will put this on hold for now and come back to it in January

smbea commented:
It appears that the bug that @barmac noticed, is also present on develop so it isn't introduced by this fix. I created an issue for this: https://github.com/camunda/camunda-modeler/issues/2654

MaxTru reviewed (approved):
:wave:
